### PR TITLE
Adapting all the required methods to migrate agent-groups files to Wazuh-DB

### PR DESCRIPTION
**Pull request by pereyra-m**
|Related issue|
|---|
|#11221|

## Description

This PR adapts all the necessary methods to stop using the agent groups files and only query Wazuh-DB for this data.
Some methods were no longer necessary under this new implementation and were removed.

The new column `groups_hash` was included in the required methods, and the information in the `belongs` table and **group/groups_hash** columns was updated in every change.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation

- Memory tests for Linux
  - [x] Scan-build report

### CODE REVIEW

**Comment by pereyra-m** [comment]
## Testing

### Scan-build

No errors were reported

scan_build

### Manual tests

#### Connecting a new agent

An agent was registered with **authd** and connected to the manager ( a **never_connected** agent shouldn't be added to groups). The agent is automatically added to **default** group.

We can see in the image that:
- The belongs table was properly updated
- Both `group` and `groups_hash` columns have the right values
- There isn't a file created under `agent-groups`

test1

#### Adding/removing an agent from a group

Now, a new group `test1` was created

```
./bin/agent_groups -a -g test1-q
```

The `agent_groups` script wasn't been updated yet, so the assignation will be performed using Wazuh-DB

```
python3 /home/vagrant/wazuh-wdb_query_sql.py global 'update-agent-group {"id":001,"group":"default,test1"}'
```

We can see that:
- All the columns were properly updated
- The multi-group folder was created

test_multi

Then, if we remove the agent from that group, all the columns get updated again.

If an invalid group is present in the list, the `group` column is not updated.

### Notes

There is a race condition until **authd** saves the new agent in **global.db**, so the registration with groups can't be tested yet.

Also, if a group is removed, the `belongs` table will be updated but not the rest of the columns in the **agent** table. This will be solved by the caller.

**Comment by pereyra-m** [comment]
### Comma separate string replacement

After analyzing the code, it was determined that even that now the files are no longer being written, the comma separated string is still used internally by the daemons. 

Currently, the Wazuh-DB helper returns a JSON array and it has to be translated to this legacy format, but it could be used as an array to avoid repeated transformations. This would require the following modifications:

- The `get_agent_group()` method is called by **remoted** to determine the agent's group. If it fails to find it, tries to guess it or assigns the `default` value. This string is later stored in the **pending_data** hash table to be later used by `read_controlmsg()` 
- The `c_files()` method stores in the `m_hash` table the value from the **group** column. It is used by `c_multi_group()` to generate the **merged.mg** sums and by `send_file_toagent()` during the agent communication

Considering the work required, this could be completed in another issue.

File: src/wazuh_db/wdb.c [hunk 1]
```diff
@@ -1145,6 +1146,51 @@ cJSON * wdb_exec_stmt(sqlite3_stmt * stmt) {
     return result;
 }
 
+cJSON * wdb_exec_stmt_single_array(sqlite3_stmt * stmt) {
```
Comment (palaciosjeremias) [inline src/wazuh_db/wdb.c:1149]: I think a name like wdb_exec_stmt_single_column will be more representative.

File: src/wazuh_db/wdb.h [hunk 2]
```diff
@@ -826,6 +827,15 @@ int wdb_exec_stmt_send(sqlite3_stmt* stmt, int peer);
  */
 cJSON * wdb_exec_stmt(sqlite3_stmt * stmt);
 
+/**
+ * @brief Function to execute a SQL statement and save the result in a single JSON array without column name like:
+ *        ["string1","string2", ...]. The query should return only one column in every step.
```
Comment (palaciosjeremias) [inline src/wazuh_db/wdb.h:832]: We can improve the example like:
["column_value_1","column_value_2", ...]

File: src/wazuh_db/wdb_parser.c [hunk 3]
```diff
@@ -5173,6 +5182,25 @@ int wdb_parse_global_insert_agent_group(wdb_t * wdb, char * input, char * output
     return OS_SUCCESS;
 }
 
+int wdb_parse_global_select_group_belong(wdb_t *wdb, char *input, char *output) {
+    int agent_id = atoi(input);
+    char *out = NULL;
```
Comment (palaciosjeremias) [inline src/wazuh_db/wdb_parser.c:5187]: We can move this declaration directly to line 5196.

File: src/remoted/manager.c [hunk 4]
```diff
@@ -790,64 +778,28 @@ static void c_files()
...
-        }
-
-        if (snprintf(path, PATH_MAX + 1, GROUPS_DIR "/%s", entry->d_name) > PATH_MAX) {
-            merror("At c_files(): path too long.");
-            break;
-        }
+    int *agents_array = wdb_get_all_agents(false, NULL);
 
-        fp = fopen(path,"r");
+    for(int i = 0; agents_array && agents_array[i] != -1; i++ ) {
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/remoted/manager.c:783]: Fixed !

File: src/remoted/manager.c [hunk 5]
```diff
@@ -790,64 +778,28 @@ static void c_files()
...
+        cJSON* j_agent_info = wdb_get_agent_info(agents_array[i], NULL);
+        if(j_agent_info) {
+            char* agent_groups = cJSON_GetStringValue(cJSON_GetObjectItem(j_agent_info->child, "group"));
+            char* agent_groups_hash = cJSON_GetStringValue(cJSON_GetObjectItem(j_agent_info->child, "groups_hash"));
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/remoted/manager.c:787]: Fixed !

File: src/shared/agent_op.c [hunk 6]
```diff
@@ -272,76 +273,31 @@ int os_write_agent_info(const char *agent_name, __attribute__((unused)) const ch
...
-
-    if (snprintf(path, PATH_MAX, GROUPS_DIR "/%s", id) >= PATH_MAX) {
-        merror("At set_agent_group(): file path too large for agent '%s'.", id);
-        return -1;
-    }
-
-    oldmask = umask(0006);
-    fp = fopen(path, "w");
-    umask(oldmask);
+int get_agent_group(int id, char *group, size_t size) {
```
[the remark this answers is not in the document]
Comment (pereyra-m) [reply src/shared/agent_op.c:276]: You are right, the method was written before the confirmation of the **group** column.
I've updated the method and also added the `wdb_sock` parameter.

File: src/wazuh_db/wdb.c (hunk 1 above)
Comment (pereyra-m) [reply src/wazuh_db/wdb.c:1149]: You are right!
Fixed!

File: src/wazuh_db/wdb.h (hunk 2 above)
Comment (pereyra-m) [reply src/wazuh_db/wdb.h:832]: Done!

File: src/wazuh_db/wdb_parser.c (hunk 3 above)
Comment (pereyra-m) [reply src/wazuh_db/wdb_parser.c:5187]: Done !

**Review by palaciosjeremias — APPROVED** [review]
LGTM

**Pull request merged**